File: core/core-api/src/main/java/com/coffee/gu/coupon/IssuedCouponService.java
package com.coffee.gu.coupon;

import com.coffee.gu.Principal;
import com.coffee.gu.menu.Menu;
import com.coffee.gu.menu.MenuFinder;
import com.coffee.gu.order.Order;
import com.coffee.gu.order.OrderLine;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class IssuedCouponService {

    private final IssuedCouponFinder issuedCouponFinder;
    private final CouponFinder couponFinder;
    private final MenuFinder menuFinder;

    public IssuedCouponService(IssuedCouponFinder issuedCouponFinder, CouponFinder couponFinder, MenuFinder menuFinder) {
        this.issuedCouponFinder = issuedCouponFinder;
        this.couponFinder = couponFinder;
        this.menuFinder = menuFinder;
    }

    public List<IssuedCoupon> getIssuedCoupons(Principal principal) {
        return issuedCouponFinder.findAllByPrincipalKey(principal.getKey());
    }

    public List<IssuedCoupon> getIssuedCouponsForCheckout(Principal principal, Order order) {
        if (order.getLines().isEmpty()) return List.of();
        List<Menu> menus = menuFinder.findAllByIdIn(order.getLines().stream().map(OrderLine::getMenuId).toList());
        List<Coupon> applicableCoupons = couponFinder.findApplicableCoupons(menus);
        if (applicableCoupons.isEmpty()) return List.of();
        return issuedCouponFinder.getUsableAllByPrincipalAndCoupons(principal, applicableCoupons);
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/coupon/IssuedCouponService.kt
package com.coffee.gu.coupon

import com.coffee.gu.Principal
import com.coffee.gu.menu.MenuFinder
import com.coffee.gu.order.Order
import org.springframework.stereotype.Service

@Service
class IssuedCouponService(
    private val issuedCouponFinder: IssuedCouponFinder,
    private val couponFinder: CouponFinder,
    private val menuFinder: MenuFinder
) {
    fun getIssuedCoupons(principal: Principal): List<IssuedCoupon> {
        return issuedCouponFinder.findAllByPrincipalKey(principal.key)
    }

    fun getIssuedCouponsForCheckout(principal: Principal, order: Order): List<IssuedCoupon> {
        if (order.lines.isEmpty()) return emptyList()
        val menus = menuFinder.findAllByIdIn(order.lines.map { it.menuId })
        val applicableCoupons = couponFinder.findApplicableCoupons(menus)
        if (applicableCoupons.isEmpty()) return emptyList()
        return issuedCouponFinder.getUsableAllByPrincipalAndCoupons(principal, applicableCoupons)
    }
}
